src/opnsense/mvc/app/controllers/OPNsense/Routes/Api/RoutesController.php

```diff
@@ -0,0 +1,169 @@
+<?php
+/**
+ *    Copyright (C) 2015 Deciso B.V.
+ *    Copyright (C) 2017 Fabian Franz
+ *
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *
+ *    1. Redistributions of source code must retain the above copyright notice,
+ *       this list of conditions and the following disclaimer.
+ *
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *
+ *    THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *    INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ *    AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *    AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ *    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+namespace OPNsense\Routes\Api;
+
+use \OPNsense\Base\ApiControllerBase;
+use \OPNsense\Core\Backend;
+use \OPNsense\Core\Config;
+use \OPNsense\Base\UIModelGrid;
+use \OPNsense\Routes\Route;
+
+/**
+ * @package OPNsense\Routes
+ */
+class RoutesController extends ApiControllerBase
+{
+    public function searchrouteAction()
+    {
+        $this->sessionClose();
+        $mdlRoute = new Route();
+        $grid = new UIModelGrid($mdlRoute->route);
+        return $grid->fetchBindRequest(
+            $this->request,
+            array("disabled", "network", "gateway", "descr"),
+            "description"
+        );
+    }
+
+    public function setrouteAction($uuid)
+    {
+        $result = array("result"=>"failed");
+        if ($this->request->isPost() && $this->request->hasPost("route")) {
+            $mdlRoute = new Route();
+            if ($uuid != null) {
+                $node = $mdlRoute->getNodeByReference('route.'.$uuid);
+                if ($node != null) {
+                    $this->backend_execute_route('delete', $node);
+                    $node->setNodes($this->request->getPost("route"));
+                    $validations = $mdlRoute->validate($node->__reference, "route");
+                    if (count($validations)) {
+                        $result['validations'] = $validations;
+                    } else {
+                        // serialize model to config and save
+                        $mdlRoute->serializeToConfig();
+                        $this->backend_execute_route('add', $node);
```

Comment:
@fabianfrz If we can't distinguish our routes with the ones created by dhcpd and/or quagga your definitely right on that point. I'm not sure what the best option is here, but apply immediate could lead to unexpected results (loss of access to the gui for example).

I need to think some more about this, ideally it would save a route configuration which we can load, but that means we need to track our changes somehow (without adding too much logic).

I don't mind if we first work on a version that just works using the current approach and refactor some in devel later on.